mkenv/converge.py

```diff
@@ -69,7 +69,10 @@ def main(filesystem, locator, link_dir, handle_error):
             )
 
         with filesystem.open(existing_config_path, "w") as existing_config:
-            existing_config.write(pytoml.dumps(config).encode("utf-8"))
+            value = pytoml.dumps(config)
```

Comment:
Isn't it easier here to just always open the file in text mode and not need to worry about the different defaults?

Reply:
I will recheck the error when I get back to my computer but I was getting something about Unicode objects not supporting the buffer interface when trying to write with `f.writelines()` in py2.  Also, isn't it already being opened in text mode?  Perhaps this ends up pointing back at me messing something up elsewhere.

Reply:
Without specifying, `"w"` is str (bytes) on Py2 and str (unicode) on Py3.

PyTOML always gives you unicode (I believe?), which is why you need to convert here on Py2. But if you opened the file in text mode (`"wt"`), you can write unicode to it on both Py2 and Py3.

Reply:
(You'll have to use `io.open` to do that by the way, because IIRC the default encoding on Py2 for `open` will be ASCII, so it will blow up for non-ASCII unicode)

Lemme know if any of that is unclear (or if anything you see contradicts it, I wrote it all from memory)

Reply:
I'm going to recheck what the error is when I get a chance. Also, I'll pull a branch for py3 dev so we can use filesystems from git until we get this settled. Then errors and fixes will be present in the public builds.

As to using 't' mode, that's a feature of `io.open()` only.  Presently, iirc, at least the memory filesystem doesn't support 't', so that would be an addition.  If my memory serves, it would be quite simple. I'm not sure about native.

Reply:
Ok, so the error in py3 was the straightforward:
```python
TypeError: write() argument must be str, not bytes
```

Did you want to add 't' mode to `filesystems`?  If I read correctly, py2 `open()` only has binary and native and (py3 `open()`, py2/3 `io.open()`) only have binary and text.  If that is correct then we would need either two open calls or a new set of mode options defined that included 'n' or such for native.

For quick reference:
* [py2 `open()`](https://docs.python.org/2/library/functions.html#open)
* [py3 `open()`](https://docs.python.org/3/library/functions.html#open)
* [py2 `io.open()`](https://docs.python.org/2/library/io.html#io.open)
* [py3 `io.open()`](https://docs.python.org/3/library/io.html#io.open) (an alias for py3 `open()`)

Reply:
Ech right, so I think I remembered the annoying part here.

So yeah I think adding a `t` mode in filesystems is the right thing to do in order to allow users (e.g. this library) to say "open a file and I want text", as opposed to getting `str` *but*, I think the default should be native strings (i.e. the same as `open`, not `io.open`), so the internals of `native.FS.open` just need to pick the right `open` depending on the mode that's passed in.

Longer term, I'd like to get rid of all this to be honest -- my first thought says that an API like `FS.open` (always) returning a `BetterFile` object which you can call `.bytes` or `.text` on to return a wrapped file object (via the `io` wrapper objects) I suspect means that all of the modes (`b`, `t`, native) can go away in favor of just using modes for read write append, and Python types can be method calls instead. But yeah for now it strikes me to just do the small thing above. How's that sound to you?

Reply:
Alrighty.  I had started to write up a change to use `io.open()` and modes of the form `[rwa][bnt]`.  I'm fine with `rn` being the default.  I guess there should be `+` too?  I don't know if it's good but I'm tempted to totally isolated the modes from the underlying modes. Not offering 'U' or whatever else without explicit consideration and consistency across environments. Thoughts?

Reply:
> I guess there should be + too?

Yeah, but basically anything that there doesn't exist a test for I wouldn't want to add until there was one for.

> I'm fine with rn being the default.

Just to be sure -- there's no such thing as `n` right? I mean, open basically silently eats modes it doesn't understand, but the actual way you get native strings is to just not specify either `t` or `b` -- I wouldn't want to invent a new character in this mode language (because someone can come along and add that new character to mean something else, and because this whole DSL is super cryptic).

> I don't know if it's good but I'm tempted to totally isolate the incoming modes from the underlying modes. Not offering 'U' or whatever else without explicit consideration and consistency across environments.

I *think* here you're agreeing with the above about not adding things until we consider each one individually right? If so definitely +1.

Reply:
So b gets you bytes and t gets you Unicode and not specifying either gets you native? No way to explicitly state native?

Reply:
Yeah -- that's how `open` works right? (Other than it lacks `t`)

Reply:
I thought an explicit way to get the default behavior might be better than open. But, no biggy. I think I no where to go with this.